### PR TITLE
Devrel 1202/files validation

### PR DESCRIPTION
**Pull request by IvanKiral**
### Motivation

Which issue does this fix? Fixes #`issue number`

If no issue exists, what is the fix or new feature? Were there any reasons to fix/implement things that are not obvious?

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

File: src/modules/sync/validation/elementSchemas.ts [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+  MultipleChoiceElementDataSchema,
+  RichTextElementDataSchema,
+  TaxonomyElementDataSchema,
+  TextElementDataSchema,
+] as const;
+
+const extendTuples = <Tuple1 extends ReadonlyArray<unknown>, Tuple2 extends ReadonlyArray<unknown>>(
+  t1: Tuple1,
+  t2: Tuple2,
+): CombineTuples<Tuple1, Tuple2> => [...t1, ...t2] as unknown as CombineTuples<Tuple1, Tuple2>;
```
Comment (JiriLojda) [inline src/modules/sync/validation/elementSchemas.ts:302]: The casting is not necessary and in fact, the whole `extendTuples` is not needed. You can do the plain `[...SnippetElementsSchemas, SnipetElementDataSchema, ...] as const` and it works the same.

File: src/modules/sync/validation/elementSchemas.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+const TypeElementSchemas = extendTuples(
+  SnippetElementsSchemas,
+  [SnippetElementDataSchema, SubpagesElementDataSchema, UrlSlugElementDataSchema] as const,
+);
+
+export const TypeElementSchemasWithGroups = TypeElementSchemas
+  .map(schema =>
+    schema.extend({ content_group: z.strictObject({ codename: z.string() }) })
```
Comment (JiriLojda) [inline src/modules/sync/validation/elementSchemas.ts:311]: Why don't you use the `ContentGroupSchema` here?

File: src/modules/sync/validation/elementSchemas.ts [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+  ) as unknown as AddPropToObjectTuple<
+    typeof TypeElementSchemas,
+    typeof ContentGroupSchema
+  >;
+
+const ContentGroupSchema = z.strictObject({
+  content_group: z.strictObject({ codename: z.string() }),
```
Comment (JiriLojda) [inline src/modules/sync/validation/elementSchemas.ts:318]: Why don't you use the `CodenameReferenceSchema` from `commonSchemas.ts`?

File: src/modules/sync/validation/entitySchema.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  codename: z.string(),
+  elements: z.array(SnippetElementsSchemasUnion),
+});
+
+const ContentGroupSchema = z.strictObject({ codename: z.string(), name: z.string() });
+
+export const TypeSchema: z.ZodType<
+  ContentTypeSyncModel,
+  z.ZodTypeDef,
+  { content_groups: ReadonlyArray<Replace<ContentTypeContracts.IContentTypeGroup, { codename: string }>> }
```
Comment (JiriLojda) [inline src/modules/sync/validation/entitySchema.ts:46]:
```suggestion
  Pick<ContentTypeSyncModel, "content_groups">
```

File: src/modules/sync/validation/entitySchema.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      name: z.string(),
+      codename: z.string(),
+      is_active: z.boolean(),
+      is_default: z.literal(false),
+      fallback_language: CodenameReferenceSchema,
+    },
+  ),
+]);
+
+const WorkflowColorSchema = z.union([
```
Comment (JiriLojda) [inline src/modules/sync/validation/entitySchema.ts:114]: Don't forget to use explicit types for the schemas as it is easy to make a mistake here, especially in this union. If you forget a a color, the TS won't complain even when you use an explicit type for the constant. The best solution is probably to have an explicitly typed object with keys being the colors and values being the appropriate schema and then generate the union from the object with `Object.values`. This way TS would check that you don't forget a color.

File: src/modules/sync/validation/entitySchema.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  z.literal("orange"),
+  z.literal("brown"),
+]);
+
+const WorkflowPublishedStepSchema = z.strictObject({
+  name: z.string(),
+  codename: z.string(),
+  create_new_version_role_ids: z.array(z.string()).length(0),
+  unpublish_role_ids: z.array(z.string()).length(0),
+});
```
Comment (JiriLojda) [inline src/modules/sync/validation/entitySchema.ts:138]: Use `satisfies` in all the `strictObject` schemas with all the properties set to required as that's the best way to ensure we don't forget even an optional property.

File: tests/unit/syncModel/validation/languageSchema.test.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      codename: null,
+      is_active: null,
+      is_default: true,
+    };
+
+    const result = LanguageSchema.safeParse(invalidData);
+    expect(result.success).toBe(false);
+  });
+
+  test("Invalid object that is empty", () => {
```
Comment (JiriLojda) [inline tests/unit/syncModel/validation/languageSchema.test.ts:129]: We use `describe/it` functions in this project. Although `test` and `it` behave the same, I would stick to only one of them.

File: tests/unit/syncModel/validation/languageSchema.test.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      fallback_language: {
+        codename: "en",
+      },
+    };
+
+    const result = LanguageSchema.safeParse(invalidData);
+    expect(result.success).toBe(false);
+  });
+
+  test("Invalid object with is_default false and missing fallback_language", () => {
```
Comment (JiriLojda) [inline tests/unit/syncModel/validation/languageSchema.test.ts:48]: Although I like the enthusiasm with tests, I would rely on TS forcing us to create the schemas properly and then `zod` to behave properly with the right schemas. You didn't write tests on all the schemas anyway. I think that one valid and one invalid test per schema is more than enough and instead we can cover all the schemas.

File: src/modules/sync/validation/entitySchema.ts (hunk 5 above)
Comment (JiriLojda) [reply src/modules/sync/validation/entitySchema.ts:114]: What is the difference between `z.enum` and `z.union` filled with `z.literal`. I noticed you used both in different places. Can we use the `z.enum` here as well? It seems to be simpler.

**Review by JiriLojda — CHANGES_REQUESTED** [review]

File: src/utils/types.ts [hunk 9]
```diff
@@ -67,3 +67,71 @@ export type AnyOnePropertyOf<Obj extends object> = [keyof Obj, keyof Obj] extend
...
+ * type Invalid = CombineTuples<[], ReadonlyArray<number>>;
+ * // Result: "never"
+ *
+ * @description
+ * - If `Tuple1` or `Tuple2` extends `ReadonlyArray<unknown>`, meaning they are not fixed-length tuples,
+ *   the resulting type is `"never"`.
+ */
+
+export type CombineTuples<Tuple1 extends ReadonlyArray<unknown>, Tuple2 extends ReadonlyArray<unknown>> =
+  ReadonlyArray<unknown> extends Tuple1 ? "never" : ReadonlyArray<unknown> extends Tuple2 ? "never"
```
[the remark this answers is not in the document]
Comment (IvanKiral) [reply src/utils/types.ts:90]: after doing other comments the type is not necessary so i removed it

File: src/modules/sync/validation/elementSchemas.ts (hunk 2 above)
Comment (IvanKiral) [reply src/modules/sync/validation/elementSchemas.ts:311]: because sadly extend does take raw shape

Comment (IvanKiral) [reply src/modules/sync/validation/elementSchemas.ts:311]: i actually used ContentGroupSchema only for type

**Review by JiriLojda — APPROVED** [review]

**Pull request merged**